src/server/pps/server/api_server.go

```diff
@@ -2391,150 +2467,156 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 			pipelinePtr     pps.EtcdPipelineInfo
 			oldPipelineInfo *pps.PipelineInfo
 		)
-		if _, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-			// Read existing PipelineInfo from PFS output repo
-			return a.pipelines.ReadWrite(stm).Update(pipelineName, &pipelinePtr, func() error {
-				var err error
+		// Read existing PipelineInfo from PFS output repo
+		if err := a.pipelines.ReadWrite(txnCtx.Stm).Update(pipelineName, &pipelinePtr, func() error {
+			var err error
 
-				// We can't recover from an incomplete pipeline info here because
-				// modifying the spec repo depends on being able to access the previous
-				// commit. We therefore use `GetPipelineInfo` which will error if the
-				// spec commit isn't working.
-				oldPipelineInfo, err = ppsutil.GetPipelineInfo(pachClient, pipelineName, &pipelinePtr)
-				if err != nil {
-					return err
-				}
+			// We can't recover from an incomplete pipeline info here because
+			// modifying the spec repo depends on being able to access the previous
+			// commit. We therefore use `GetPipelineInfo` which will error if the
+			// spec commit isn't working.
+			oldPipelineInfo, err = ppsutil.GetPipelineInfo(txnCtx.Client, pipelineName, &pipelinePtr)
+			if err != nil {
+				return err
+			}
 
-				// Cannot disable stats after it has been enabled.
-				if oldPipelineInfo.EnableStats && !pipelineInfo.EnableStats {
-					return newErrPipelineUpdate(pipelineInfo.Pipeline.Name, "cannot disable stats")
-				}
+			// Cannot disable stats after it has been enabled.
+			if oldPipelineInfo.EnableStats && !pipelineInfo.EnableStats {
+				return newErrPipelineUpdate(pipelineInfo.Pipeline.Name, "cannot disable stats")
+			}
 
-				// Modify pipelineInfo (increment Version, and *preserve Stopped* so
-				// that updating a pipeline doesn't restart it)
-				pipelineInfo.Version = oldPipelineInfo.Version + 1
-				if oldPipelineInfo.Stopped {
-					provenance = nil // CreateBranch() below shouldn't create new output
-					pipelineInfo.Stopped = true
-				}
-				if !request.Reprocess {
-					pipelineInfo.Salt = oldPipelineInfo.Salt
-				}
-				// Must create spec commit before restoring output branch provenance, so
-				// that no commits are created with a mismatched spec commit
-				specCommit, err := a.makePipelineInfoCommit(pachClient, pipelineInfo)
+			// Modify pipelineInfo (increment Version, and *preserve Stopped* so
+			// that updating a pipeline doesn't restart it)
+			pipelineInfo.Version = oldPipelineInfo.Version + 1
+			if oldPipelineInfo.Stopped {
+				provenance = nil // CreateBranch() below shouldn't create new output
+				pipelineInfo.Stopped = true
+			}
+			if !request.Reprocess {
+				pipelineInfo.Salt = oldPipelineInfo.Salt
+			}
+			specCommit, err := createOrValidateSpecCommit()
+			if err != nil {
+				return err
+			}
+
+			// move the spec branch for this pipeline to the new commit
+			txnCtx.Pfs().CreateBranchInTransaction(txnCtx, &pfs.CreateBranchRequest{
```

Comment:
This consistently errors from not being able to find the commit we just made, but ends up working (in the STM retry, maybe not the first time). I assume this means I'm accidentally creating an entry in the transaction's STM for it which prevents it from going to PFS for real, but I'm not sure.

Reply:
FWIW it's definitely resulting from the actual etcd fetch not finding the commit that was just made. I'm not entirely sure I understand the model, here - the etcd client being used in pfs through the grpc is different than the one used in the transaction, but is the underlying data store distributed, leading to an eventual-consistency situation?